File: web/src/api/features.ts
import { apiFetch, ifMatchHeader } from './client'
import type { FeatureDetail, FeaturesPage, Priority, WorkflowStatus } from './types'

export interface FeatureListFilters {
  status?: WorkflowStatus
  priority?: Priority
  creator?: string
  updatedSince?: string
}

export async function listFeatures(
  projectKey: string,
  filters: FeatureListFilters = {},
  cursor?: string,
  limit?: number,
): Promise<FeaturesPage> {
  const params = new URLSearchParams()
  if (filters.status) params.set('status', filters.status)
  if (filters.priority) params.set('priority', filters.priority)
  if (filters.creator) params.set('creator', filters.creator)
  if (filters.updatedSince) params.set('updated_since', filters.updatedSince)
  if (cursor) params.set('cursor', cursor)
  if (limit) params.set('limit', String(limit))
  const query = params.toString()
  return apiFetch<FeaturesPage>(
    `/projects/${encodeURIComponent(projectKey)}/features${query ? `?${query}` : ''}`,
  )
}

export async function getFeature(ref: string, includeDeleted = false): Promise<FeatureDetail> {
  const query = includeDeleted ? '?include_deleted=true' : ''
  return apiFetch<FeatureDetail>(`/features/${encodeURIComponent(ref)}${query}`)
}

export interface CreateFeatureInput {
  title: string
  description: string
  priority: Priority
}

export async function createFeature(
  projectKey: string,
  input: CreateFeatureInput,
): Promise<FeatureDetail> {
  return apiFetch<FeatureDetail>(`/projects/${encodeURIComponent(projectKey)}/features`, {
    method: 'POST',
    body: input,
  })
}

export interface UpdateFeatureInput {
  title: string
  description: string
  priority: Priority
}

/** PATCH /features/{ref} is a full-representation update despite the
 * verb (internal/httpapi/features.go's updateFeatureRequest has no
 * partial-update semantics) — resend every field. */
export async function updateFeature(
  ref: string,
  input: UpdateFeatureInput,
  expectedVersion: number,
): Promise<FeatureDetail> {
  return apiFetch<FeatureDetail>(`/features/${encodeURIComponent(ref)}`, {
    method: 'PATCH',
    body: input,
    headers: ifMatchHeader(expectedVersion),
  })
}

/** POST /features/{ref}/status — single-field status mutation, its
 * own endpoint (Phase 4 addition) rather than a field on
 * UpdateFeatureInput, mirroring updateTicketStatus's split from
 * updateTicketFields. */
export async function updateFeatureStatus(
  ref: string,
  status: WorkflowStatus,
  expectedVersion: number,
): Promise<FeatureDetail> {
  return apiFetch<FeatureDetail>(`/features/${encodeURIComponent(ref)}/status`, {
    method: 'POST',
    body: { status },
    headers: ifMatchHeader(expectedVersion),
  })
}

export async function reorderFeature(
  ref: string,
  afterRef: string | null,
  expectedVersion: number,
): Promise<FeatureDetail> {
  return apiFetch<FeatureDetail>(`/features/${encodeURIComponent(ref)}/reorder`, {
    method: 'POST',
    body: { after_ref: afterRef },
    headers: ifMatchHeader(expectedVersion),
  })
}

export async function deleteFeature(
  ref: string,
  expectedVersion: number,
  cascade = false,
): Promise<{ version: number }> {
  const query = cascade ? '?cascade=true' : ''
  return apiFetch<{ version: number }>(`/features/${encodeURIComponent(ref)}${query}`, {
    method: 'DELETE',
    headers: ifMatchHeader(expectedVersion),
  })
}

export async function restoreFeature(ref: string, expectedVersion: number): Promise<FeatureDetail> {
  return apiFetch<FeatureDetail>(`/features/${encodeURIComponent(ref)}/restore`, {
    method: 'POST',
    headers: ifMatchHeader(expectedVersion),
  })
}

File: internal/apiclient/features.go
package apiclient

import (
	"context"
	"net/http"
	"net/url"
	"time"
)

// Feature mirrors internal/httpapi/wire.go's featureDetail
// field-for-field — see dto.go's top comment for why this isn't
// internal/domain.Feature.
type Feature struct {
	Ref         string     `json:"ref"`
	Project     string     `json:"project"`
	Title       string     `json:"title"`
	Description string     `json:"description"`
	Status      string     `json:"status"`
	Priority    string     `json:"priority"`
	Creator     *string    `json:"creator,omitempty"`
	Version     int64      `json:"version"`
	CreatedAt   time.Time  `json:"created_at"`
	UpdatedAt   time.Time  `json:"updated_at"`
	DeletedAt   *time.Time `json:"deleted_at,omitempty"`
}

// FeatureCompact mirrors internal/httpapi/wire.go's featureCompact —
// list rows never carry Description, per product spec §7.2/§11.
type FeatureCompact struct {
	Ref       string    `json:"ref"`
	Title     string    `json:"title"`
	Status    string    `json:"status"`
	Priority  string    `json:"priority"`
	Version   int64     `json:"version"`
	UpdatedAt time.Time `json:"updated_at"`
}

// FeaturesPage is GET /projects/{key}/features' response envelope.
type FeaturesPage struct {
	Features   []FeatureCompact `json:"features"`
	NextCursor string           `json:"next_cursor,omitempty"`
}

type FeatureListFilters struct {
	Status       string
	Priority     string
	Creator      string
	UpdatedSince string
}

// CreateFeatureRequest is POST /projects/{key}/features' request body.
type CreateFeatureRequest struct {
	Title       string `json:"title"`
	Description string `json:"description"`
	Priority    string `json:"priority,omitempty"`
}

// UpdateFeatureRequest is PATCH /features/{ref}'s request body — a
// full-representation update, matching internal/httpapi's
// updateFeatureRequest (no separate status-only route the way tickets
// have; features have no PUT/PATCH split).
type UpdateFeatureRequest struct {
	Title       string `json:"title"`
	Description string `json:"description"`
	Priority    string `json:"priority"`
}

// CreateFeature is POST /projects/{key}/features.
func (c *Client) CreateFeature(ctx context.Context, projectKey string, req CreateFeatureRequest, idempotencyKeys ...string) (Feature, error) {
	idempotencyKey := ""
	if len(idempotencyKeys) > 0 {
		idempotencyKey = idempotencyKeys[0]
	}
	var f Feature
	err := c.do(ctx, http.MethodPost, "/projects/"+url.PathEscape(projectKey)+"/features", req, &f, requestOptions{IdempotencyKey: idempotencyKey})
	return f, err
}

// GetFeature is GET /features/{ref}.
func (c *Client) GetFeature(ctx context.Context, ref string) (Feature, error) {
	var f Feature
	err := c.do(ctx, http.MethodGet, "/features/"+url.PathEscape(ref), nil, &f, requestOptions{})
	return f, err
}

func (c *Client) GetFeatureIncludingDeleted(ctx context.Context, ref string) (Feature, error) {
	var f Feature
	err := c.do(ctx, http.MethodGet, "/features/"+url.PathEscape(ref)+"?include_deleted=true", nil, &f, requestOptions{})
	return f, err
}

// ListFeatures is GET /projects/{key}/features. Compact rows only.

func (c *Client) ListFeatures(ctx context.Context, projectKey string, limit int, cursor string) (FeaturesPage, error) {
	return c.ListFeaturesFiltered(ctx, projectKey, FeatureListFilters{}, limit, cursor)
}

func (c *Client) ListFeaturesFiltered(ctx context.Context, projectKey string, filters FeatureListFilters, limit int, cursor string) (FeaturesPage, error) {
	var page FeaturesPage
	q := url.Values{}
	if filters.Status != "" {
		q.Set("status", filters.Status)
	}
	if filters.Priority != "" {
		q.Set("priority", filters.Priority)
	}
	if filters.Creator != "" {
		q.Set("creator", filters.Creator)
	}
	if filters.UpdatedSince != "" {
		q.Set("updated_since", filters.UpdatedSince)
	}
	path := "/projects/" + url.PathEscape(projectKey) + "/features" + listQuery(q, limit, cursor)
	err := c.do(ctx, http.MethodGet, path, nil, &page, requestOptions{})
	return page, err
}

// UpdateFeature is PATCH /features/{ref} — a full-representation
// update, so (unlike UpdateTicket) there is no merge-from-unset-fields
// convenience here; the caller supplies every field.
func (c *Client) UpdateFeature(ctx context.Context, ref string, req UpdateFeatureRequest, expectedVersion int64) (Feature, error) {
	var f Feature
	err := c.do(ctx, http.MethodPatch, "/features/"+url.PathEscape(ref), req, &f, requestOptions{IfMatch: &expectedVersion})
	return f, err
}

// SetFeatureStatus is POST /features/{ref}/status — the single-field
// workflow-status mutation UpdateFeature's full-representation PATCH
// doesn't cover, mirroring updateTicketStatus's own split from
// UpdateTicket.
func (c *Client) SetFeatureStatus(ctx context.Context, ref, status string, expectedVersion int64) (Feature, error) {
	var f Feature
	err := c.do(ctx, http.MethodPost, "/features/"+url.PathEscape(ref)+"/status",
		struct {
			Status string `json:"status"`
		}{Status: status},
		&f, requestOptions{IfMatch: &expectedVersion})
	return f, err
}

// ReorderFeature is POST /features/{ref}/reorder.
func (c *Client) ReorderFeature(ctx context.Context, ref string, afterRef *string, expectedVersion int64) (Feature, error) {
	var f Feature
	err := c.do(ctx, http.MethodPost, "/features/"+url.PathEscape(ref)+"/reorder",
		struct {
			AfterRef *string `json:"after_ref"`
		}{AfterRef: afterRef},
		&f, requestOptions{IfMatch: &expectedVersion})
	return f, err
}

// DeleteFeature is DELETE /features/{ref}?cascade=true|false (soft
// delete). ADR 0013 blocks it by default when the feature still holds
// non-deleted tickets — cascade soft-deletes them too.
func (c *Client) DeleteFeature(ctx context.Context, ref string, cascade bool, expectedVersion int64) (int64, error) {
	path := "/features/" + url.PathEscape(ref)
	if cascade {
		path += "?cascade=true"
	}
	var resp struct {
		Version int64 `json:"version"`
	}
	err := c.do(ctx, http.MethodDelete, path, nil, &resp, requestOptions{IfMatch: &expectedVersion})
	return resp.Version, err
}

// RestoreFeature is POST /features/{ref}/restore.
func (c *Client) RestoreFeature(ctx context.Context, ref string, expectedVersion int64) (Feature, error) {
	var f Feature
	err := c.do(ctx, http.MethodPost, "/features/"+url.PathEscape(ref)+"/restore", nil, &f, requestOptions{IfMatch: &expectedVersion})
	return f, err
}
